service/udp.go

```diff
@@ -254,16 +254,150 @@ func (h *packetHandler) validatePacket(textData []byte) ([]byte, *net.UDPAddr, *
 	return payload, tgtUDPAddr, nil
 }
 
+type AssociationHandleFunc func(ctx context.Context, conn net.Conn)
+
+// PacketServe listens for UDP packets on the provided [net.PacketConn] and creates
+// and manages NAT associations. It uses a NAT map to track active associations and
+// handles their lifecycle.
+func PacketServe(clientConn net.PacketConn, assocHandle AssociationHandleFunc, metrics NATMetrics) {
+	nm := newNATmap()
+	ctx, contextCancel := context.WithCancel(context.Background())
+	defer contextCancel()
+
+	for {
+		lazySlice := readBufPool.LazySlice()
+		buffer := lazySlice.Acquire()
+
+		expired := false
+		func() {
+			defer func() {
+				if r := recover(); r != nil {
+					slog.Error("Panic in UDP loop. Continuing to listen.", "err", r)
+					debug.PrintStack()
+					lazySlice.Release()
+				}
+			}()
+			n, addr, err := clientConn.ReadFrom(buffer)
+			if err != nil {
+				lazySlice.Release()
+				if errors.Is(err, net.ErrClosed) {
+					expired = true
+					return
+				}
+				slog.Warn("Failed to read from client. Continuing to listen.", "err", err)
+				return
+			}
+			pkt := &packet{payload: buffer[:n], done: lazySlice.Release}
+
+			// TODO(#19): Include server address in the NAT key as well.
+			assoc := nm.Get(addr.String())
+			if assoc == nil {
+				assoc = &association{
+					pc:     clientConn,
+					raddr:  addr,
```

Comment:
`raddr` doesn't communicate what it is.

```suggestion
					clientAddr:  clientAddr,
```